.github/workflows/celerity_ci.yml

```
@@ -70,96 +92,58 @@ jobs:
             --config_file=${{ env.container-workspace }}/.clang-tidy \
             --include="*.h,*.cc,*.[ch]pp"
 
-  build-and-test:
+  # We need to jump through some hoops to have different build matrices based on what triggered the workflow.
+  # For normal CI runs we want to build and test against everything except the "HEAD" revisions, whereas during
+  # nightly builds we *only* want those.
+  #
+  # Current workaround is to represent the matrix as a JSON object, which is then deserialized in the next job.
+  read-build-matrix:
     needs: find-duplicate-workflows
+    if: needs.find-duplicate-workflows.outputs.should_skip != 'true'
+    runs-on: self-hosted
+    outputs:
+      matrix: ${{ steps.finalize-json-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v2
+      - id: finalize-json-matrix
+        name: Read build matrix from file
+        shell: python
+        run: |
+          import json
+          with open("${{ github.workspace }}/.github/workflows/build_matrix.json") as f:
+            matrices = json.load(f)
+            true_false = { 'true': True, 'false': False } # Hack for getting YAML booleans into python
+            is_regular_run = true_false['${{ github.event_name != 'schedule' && inputs.test-head == false }}']
+            default_matrix = json.dumps(matrices['default' if is_regular_run else 'nightly'])
+            print('::set-output name=matrix::{ "include":%s }' % default_matrix)
```

This script could be a bit less convoluted:

```suggestion
          if '${{ github.event_name != 'schedule' && inputs.test-head == false }}' == 'true':
            matrix = matrices['default']
          else:
            matrix = matrices['nightly']
          endif
          print('::set-output name=matrix::{ "include":%s }' % json.dumps(matrix))
```

---

Thanks, done!